src/Microsoft.AspNet.Authentication/AuthenticationHandler.cs

```diff
@@ -296,27 +314,37 @@ protected virtual Task<bool> HandleUnauthorizedAsync(ChallengeContext context)
                 {
                     case ChallengeBehavior.Automatic:
                         // If there is a principal already, invoke the forbidden code path
-                        var ticket = await HandleAuthenticateOnceAsync();
-                        if (ticket?.Principal != null)
+                        var result = await HandleAuthenticateOnceAsync();
+
+                        //if (result?.Error != null)
+                        //{
+                        //    await HandleErrorAsync(result.Error);
+                        //    if (result.Error.IsRequestComplete)
+                        //    {
+                        //        context.CompleteRequest();
+                        //    }
+                        //}
+
+                        if (result?.Ticket?.Principal != null)
                         {
-                            handled = await HandleForbiddenAsync(context);
+                            await HandleForbiddenAsync(context);
                         }
                         else
                         {
-                            handled = await HandleUnauthorizedAsync(context);
+                            await HandleUnauthorizedAsync(context);
                         }
                         break;
                     case ChallengeBehavior.Unauthorized:
-                        handled = await HandleUnauthorizedAsync(context);
+                        await HandleUnauthorizedAsync(context);
                         break;
                     case ChallengeBehavior.Forbidden:
-                        handled = await HandleForbiddenAsync(context);
+                        await HandleForbiddenAsync(context);
                         break;
                 }
                 context.Accept();
             }
 
-            if (!handled && PriorHandler != null)
+            if (PriorHandler != null)
```

Comment:
Blah, no this is a regression, we used to be able to rely on the Control flow in ChallengeContext, but we reverted those.  So our options are to bring control flow back in ChallengeContext for this, or go with our sort of lame book return again.  I prefer putting the control flow into ChallengeContext and checking that.  Thoughts?